pkg/controller/job/job_controller.go

```diff
@@ -1377,6 +1398,12 @@ func (jm *Controller) enactJobFinished(logger klog.Logger, jobCtx *syncJobCtx) b
 			return false
 		}
 	}
+	if delayTerminalCondition() {
+		if *jobCtx.terminating > 0 {
+			logger.V(4).Info("Delaying marking the Job as finished, because there are still terminating pod(s)", "job", klog.KObj(job), "condition", jobCtx.finishedCondition.Type, "count", *jobCtx.terminating)
+			return false
+		}
+	}
```

Comment:
I was rather thinking of ensuring this at the validation level, that when setting Terminal condition, `active=0, ready<=active, and terminating=0`, effectively re-adding the rules as in: https://github.com/kubernetes/kubernetes/pull/123792. when JobManagedBy is enabled. The advantage of validation is that it will ensure conformance for other Job controller implementations. WDYT?

Reply:
Well, bugs can happen also when feature gate is GA, so we should think about it more.

I think these are main scenarios:
1. there is a bug in the Job controller which prevents adding terminal condition even though all pods are terminal, and don't have finalizers
2. there is a bug in Kubelet (see [here](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase)
3. there is a bug in PodGC to transition it to Failed phase

For (1.) I believe we could just fix inside the controller, as with any other bug.
For (2.) or (3.) we may not have means to mitigate inside Job controller other than disabling the FG (but only until GA). 

I guess we need to decide if these (2.) and (3) cases are critical for the workflows. I would argue the consequence is rather limited, I think it only affects CronJobs which could prevent new job creation when concurrencyPolicy=Forbid. It can also affect external in-house controllers but they could adjust by using `failed || FailureTarget`.

Reply:
I'm ok with what is implemented here and the API validation updates.

Reply:
ack, api validation should be enough

Reply:
It is not clear to me whether the resolution of this discussion was: "add additional validation that terminal conditions can't be set when `active > 0 || ready > active || terminating > 0`" or "no additional change is needed over what the PR currently implements"?

Reply:
My reading of the resolution, which is currently implemented in was, add the following validation rules:
1. ready <= active (new rule in https://github.com/kubernetes/kubernetes/pull/125849 , see [here](https://github.com/kubernetes/kubernetes/blob/3c568d2d527f79b269f9433ebf9c6937a43d97cd/pkg/apis/batch/validation/validation.go#L586-L590))
2. terminal conditions can't be set when terminating > 0 (new rule in https://github.com/kubernetes/kubernetes/pull/125849 , see [here](https://github.com/kubernetes/kubernetes/blob/3c568d2d527f79b269f9433ebf9c6937a43d97cd/pkg/apis/batch/validation/validation.go#L581-L584))
3. terminal conditions can't be set when active > 0 (already on master, see [here](https://github.com/kubernetes/kubernetes/blob/27c6c30905107593845d4b31281f03fc8b39cc9f/pkg/apis/batch/validation/validation.go#L523-L525)

From 1 and 3 it follows also that terminal conditions can't be set when `ready>0`, so it does not need an explicit rule.

Reply:
After recent changes this is now part of this PR, and that was my understanding as well.